timed/authentication.py

```diff
@@ -0,0 +1,109 @@
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def get_userinfo_or_introspection(self, access_token):
+        try:
+            claims = self.cached_request(
+                self.get_userinfo, access_token, "auth.userinfo"
+            )
+        except requests.HTTPError as e:
+            if not (
+                e.response.status_code in [401, 403] and settings.OIDC_CHECK_INTROSPECT
+            ):
+                raise e
+
+            # check introspection if userinfo fails (confidental client)
+            claims = self.cached_request(
+                self.get_introspection, access_token, "auth.introspection"
+            )
+            if "client_id" not in claims:
+                raise SuspiciousOperation("client_id not present in introspection")
+
+        return claims
+
+    def get_or_create_user(self, access_token, id_token, payload):
+        """Verify claims and return user, otherwise raise an Exception."""
+
+        claims = self.get_userinfo_or_introspection(access_token)
+
+        # claims_verified = self.verify_claims(userinfo) or self.verify_claims(introspection)
+
+        # if not claims_verified:
+        #     msg = "Claims verification failed"
+        #     raise SuspiciousOperation(msg)
+
+        users = self.filter_users_by_claims(claims)
+
+        if len(users) == 1:
+            return self.update_user(users[0], claims)
+        elif settings.OIDC_CREATE_USER:
+            return self.create_user(claims)
+        else:
+            LOGGER.debug(
+                "Login failed: No user with username %s found, and "
+                "OIDC_CREATE_USER is False",
+                self.get_username(claims),
+            )
+            return None
+
+    def filter_users_by_claims(self, claims):
+        username = self.get_username(claims)
+        return self.UserModel.objects.filter(username=username)
+
+    def cached_request(self, method, token, cache_prefix):
+        token_hash = hashlib.sha256(force_bytes(token)).hexdigest()
+
+        func = functools.partial(method, token, None, None)
+
+        return cache.get_or_set(
+            f"{cache_prefix}.{token_hash}",
+            func,
+            timeout=settings.OIDC_BEARER_TOKEN_REVALIDATION_TIME,
+        )
+
+    def create_user(self, claims):
+        """Return object for a newly created user account."""
+        username = self.get_username(claims)
+        email = claims.get(settings.OIDC_EMAIL_CLAIM, "")
+        first_name = claims.get(settings.OIDC_FIRSTNAME_CLAIM, "")
+        last_name = claims.get(settings.OIDC_LASTNAME_CLAIM, "")
+        return self.UserModel.objects.create(
+            username=username, email=email, first_name=first_name, last_name=last_name
+        )
+
+    def get_username(self, claims):
+        try:
+            return claims[settings.OIDC_USERNAME_CLAIM]
+        except KeyError:
+            raise SuspiciousOperation("Couldn't find username claim")
```

Comment:
Did you intentionally not add an `update()` method?

Reply:
Yes, this is not a requirement for now. I left `create_user()` in there just in case (it's behind the settings flag), but the normal workflow is for admins to create the users manually

Reply:
FYI I removed the call to `self.update_user()` because it's a noop anyways.